### PR TITLE
Adds in Tanks as vehicles

### PR DESCRIPTION
**Pull request by Draconic-saint**
The tank vehicle: Standard it moves at the speed of a regular human and has it's firing mechanism disabled. Provides no additional armor or protection from harm.

Current version has a gyro-jet round as "shell". Movement is disabled for one second after firing. This is because otherwise the driver would be able to drive into their own round after firing it, effectively shooting themselves.

It'll only exist in the code for now, no way to acquire on station unless the admins make it so.

:cl: Jedr
Adds in the tank, an admin-only vehicle that can shoot explosive rounds.
/:cl:

### CODE REVIEW

**Comment by CrAzYPiLoT-SS13** [comment]
You could attribute the blocked movement as recoil. Also, pix pl0x

**Comment by Draconic-saint** [comment]
http://i.imgur.com/hYqHOnv.gifv

 Gif of tank one

http://i.imgur.com/Q6U5usC.png

Green vs Grey


In the vars I did make the blocked movement recoil_delay

File: code/modules/vehicle/tank.dm [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	icon_state = "tank_green"
+	vehicle_move_delay = 4
+	var/allow_fire = 0
+	var/recoil_delay = 10
+	var/resume_move = 10
+	var/datum/action/innate/fire_cannon/fire_cannon_action = new
+
+
+/obj/vehicle/tank/New()
+	..()
```
Comment (Fox-McCloud) [inline code/modules/vehicle/tank.dm]: Don't define something like this if it does nothing, you're just adding to overhead on creation for no reason at all.

File: code/modules/vehicle/tank.dm [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	var/bullet_type = /obj/item/projectile/bullet/gyro
+	var/turf/curloc = get_turf(target)
+	var/fire_delay = 50
+
+	if(next_firetime > world.time)
+		to_chat(owner, "<span class='warning'>Your weapons are recharging.</span>")
+
+	else
+		var/obj/item/projectile/A = new bullet_type(curloc)
+		A.dumbfire(owner.dir)
```
Comment (Fox-McCloud) [inline code/modules/vehicle/tank.dm]: outdated proc used only by spacepods.

File: code/modules/vehicle/tank.dm [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	else
+		..()
+
+/obj/vehicle/tank/buckle_mob(mob/living/M)
+	..()
+	GrantActions(buckled_mob)
+
+
+/obj/vehicle/tank/unbuckle_mob(mob/living/M)
+	fire_cannon_action.Remove(buckled_mob)
```
Comment (Fox-McCloud) [inline code/modules/vehicle/tank.dm]: I bet you have issues if the vehicle gets destroyed but not the mob on it.

**Comment by Fox-McCloud** [comment]
The sprites for this are pretty...well, not great, IMO. They're flat, and don't really fit in with the artistic style, at all.

**Comment by Draconic-saint** [comment]
On topic of the sprites, I assume you in particulary mean the North/South sprites, since those usually have more of an angle, while East/West sprites of most vehicles do usually have a more direct side-view. 

Correct me if i'm wrong though, it's just the trend that i've noticed.

File: code/modules/vehicle/tank.dm [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+		if(target.dir == EAST)
+			A.Angle = 90
+
+		if(target.dir == SOUTH)
+			A.Angle = 180
+
+		if(target.dir == WEST)
+			A.Angle = 270
+		A.fire()
```
Comment (Fox-McCloud) [inline code/modules/vehicle/tank.dm]: I thinkkkkk the way the experimentor handles the fireball may be more applicable here.

File: code/modules/vehicle/tank.dm [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	var/bullet_type = /obj/item/projectile/bullet/gyro
+	var/turf/curloc = get_turf(target)
+	var/fire_delay = 50
+
+	if(next_firetime > world.time)
+		to_chat(owner, "<span class='warning'>Your weapons are recharging.</span>")
+
+	else
+		var/obj/item/projectile/A = new bullet_type(curloc)
+		if(target.dir == NORTH)
```
Comment (Fox-McCloud) [inline code/modules/vehicle/tank.dm]: this can all be simplified with the `dir2angle` proc.

**Comment by Iamgoofball** [comment]
@Fox-McCloud just add some Army Men sprites and throw 'em in with those, put them on an away mission, call it a day

File: code/modules/vehicle/tank.dm [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	icon = 'icons/vehicles/tanks.dmi'
+	icon_state = "tank_green"
+	vehicle_move_delay = 3
+	var/allow_fire = 0
+	var/recoil_delay = 10
+	var/resume_move = 10
+	var/datum/action/innate/fire_cannon/fire_cannon_action = new
+
+/obj/vehicle/tank/handle_vehicle_layer()
+	layer = MOB_LAYER+0.1
```
Comment (tigercat2000) [inline code/modules/vehicle/tank.dm]: spacing on operators please `MOB_LAYER + 0.1`

File: code/modules/vehicle/tank.dm [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+
+/obj/vehicle/tank/proc/GrantActions(mob/living/carbon/buckled_mob)
+	if(allow_fire == 1)
```
Comment (tigercat2000) [inline code/modules/vehicle/tank.dm]: don't do `== 1`, just `if(allow_fire)`

File: code/modules/vehicle/tank.dm [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+/datum/action/innate/fire_cannon/Activate()
+	var/obj/vehicle/tank/T = target
+	var/bullet_type = /obj/item/projectile/bullet/gyro
+	var/turf/curloc = get_turf(target)
+	var/fire_delay = 50
+
+	if(next_firetime > world.time)
+		to_chat(owner, "<span class='warning'>Your weapons are recharging.</span>")
+
+	else
```
Comment (tigercat2000) [inline code/modules/vehicle/tank.dm]: no else, just do
```
if(next_firetime > world.time)
    to_chat(owner, "<span class='warning'>You weapons are recharging.</span>")
    return

//other code
```

File: code/modules/vehicle/tank.dm [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+		var/turf/MT = get_step(T, T.dir)
+
+		A.original = MT
+		A.current = start
+		A.yo = MT.y - start.y
+		A.xo = MT.x - start.x
+		A.fire()
+
+
+		next_firetime = world.time+fire_delay
```
Comment (tigercat2000) [inline code/modules/vehicle/tank.dm]: spacing on operators, `world.time + fire_delay`

File: code/modules/vehicle/tank.dm [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+		T.resume_move = world.time+T.recoil_delay
```
Comment (tigercat2000) [inline code/modules/vehicle/tank.dm]: spacing on operators

File: code/modules/vehicle/tank.dm [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+
+/obj/vehicle/tank/Move(atom/OldLoc, Dir)
+	if(resume_move>world.time)
```
Comment (tigercat2000) [inline code/modules/vehicle/tank.dm]: spacing on operators

**Comment by Draconic-saint** [comment]
@Fox-McCloud How's this for an improvement?
Left is the old version, right is the new one.

**Comment by Fox-McCloud** [comment]
It's better, but the sprites still look like they were made for some other game---mostly due to how they're shaded, the colors, and how flat they still look.

**Comment by Draconic-saint** [comment]
@Fox-McCloud @TheDZD 





Another version of the green tank. I'm also going to remove the grey tank for now. If someone wants to add tank subtypes, the can by all means. But having to redo more sprites is just not something I feel up to for the moment.

File: code/modules/vehicle/tank.dm [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+/obj/vehicle/tank/handle_vehicle_layer()
+	layer = MOB_LAYER + 0.1
+
+
+/obj/vehicle/tank/handle_vehicle_offsets()
+	..()
+	if(buckled_mob)
+		switch(buckled_mob.dir)
+			if(SOUTH)
```
Comment (tigercat2000) [inline code/modules/vehicle/tank.dm]: you can do `if(NORTH, SOUTH)` for multi-case switches

File: code/modules/vehicle/tank.dm [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+				buckled_mob.pixel_y = 4
+			if(NORTH)
+				buckled_mob.pixel_y = 4
+			if(EAST)
```
Comment (tigercat2000) [inline code/modules/vehicle/tank.dm]: same here

File: code/modules/vehicle/tank.dm [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+
+/obj/vehicle/tank/unbuckle_mob(mob/living/M)
+	fire_cannon_action.Remove(buckled_mob)
+	..()
+
+
+/obj/vehicle/tank/Destroy()
+	fire_cannon_action.Remove(buckled_mob)
+	..()
```
Comment (Fox-McCloud) [inline code/modules/vehicle/tank.dm]: `return ..()`

**Comment by Fox-McCloud** [comment]
May want to update your OP as its no longer accurate.

**Comment by Draconic-saint** [comment]
I don't know what else needs to be done to get this merged, so more feedback would be appreciated.

In my eyes, and several others I asked for feedback, the current sprites are in fact suitable for being added to the game. And when it comes to code I think I resolved every issue addressed here? Do correct me if i'm wrong.

**Comment by Signmark** [comment]
@Draconic-saint Hey what about Xenoarch?

**Comment by Draconic-saint** [comment]
@Signmark It's on hold untill a way is found to implement the mapping in a way to keep replayability high. CrazyLemon is currently working on a system that could help with that.

@Fox-McCloud it's been about 2 weeks, I'd really like to get some more information, please.

**Comment by Plotron** [comment]
I think these tanks feel slightly too retro-ish.
Imagine having sci-fi Hovertanks instead! Find inspiration in Command & Conquer.

**Comment by Draconic-saint** [comment]
The retro feel is on purpose. Just like we still have a TV in the detectives office, and medieval armour for the chaplain, why not WW2 style tanks?

**Comment by LightFire53** [comment]
The issues I,  as a player, see with this:

1. Size: Mini tanks. They look a little funny, honestly. 
2. In space: Why would a multimillion dollar corporation equip a science station with miniature tanks? Seeing as it sounds like a gun without taking a hand slot, there is really no reason...
3. When: when would we use these anywhere? Besides for maybe an away mission as a hostile mob, I don't see where these fit.
4: Other logistics: How would it be obtained? Does it require ammo? 

Again, just issues I see as a player. If we have a reason to add it and not just for the lols, then sure. If not, well, we have enough of the lol stuff as is.

**Comment by Fox-McCloud** [comment]
After gathering some opinions on this and talking it over with @TheDZD everyone seems to be opposed to this and feels it doesn't fit--as such, going to go ahead and close this.

**Comment by Draconic-saint** [comment]
@Fox-McCloud I'd like to know more about said opinions and how they were gathered, what aspect of them doesn't fit, etc.
Basically, debrief me proper, instead of just going:" Didn't fit, gg no re."

I understand if you don't want to disclose names, but I'd really like to know where I went wrong, and what the seemingly clashing opinions are. Because when i've talked to people, they were enthusiastic about the tiny comedy tanks.

also @LightFire53 There's plenty of admin-only oddball stuff in the game. I don't see the issue with expanding on that.

I probably should have stated FAR more EXPLICITLY, that this was supposed to be an **admin-only, shits and giggles vehicle.**

It was supposed to be a fun little drop of content, for the occasional usage by admins.
No ammo restrictions once firing was unlocked, no obtaining it without admins spawning it in.

**Comment by Fox-McCloud** [comment]
Everyone involved pretty much said the same thing, so I'll word it bluntly:

- Doesn't fit 
- Too niche to be used
- Doesn't make a whole lot of sense

Maintainers felt largely the same; "what's the point; this will never get used".

**Pull request closed without merging**